File: backend/src/scoring/scoring.ts
const GOAL_BONUS_PER_GOAL = 1;

export type PredictionForScoring = {
  predictedHomeScore: number;
  predictedAwayScore: number;
  predictedTopScorer: string | null;
};

export type MatchForScoring = {
  homeScore: number;
  awayScore: number;
  topScorerName: string | null;
  topScorerGoals: number | null;
};

export type MatchPointsBreakdown = {
  exactScorePoints: number;
  outcomePoints: number;
  topScorerPoints: number;
  goalBonusPoints: number;
  total: number;
};

function normalizePlayerName(name: string): string {
  return name.trim().toLowerCase();
}

function isCorrectOutcome(
  predictedHome: number,
  predictedAway: number,
  actualHome: number,
  actualAway: number,
): boolean {
  return (
    Math.sign(predictedHome - predictedAway) ===
    Math.sign(actualHome - actualAway)
  );
}

export function scorePrediction(
  prediction: PredictionForScoring,
  match: MatchForScoring,
): MatchPointsBreakdown {
  let exactScorePoints = 0;
  let outcomePoints = 0;
  let topScorerPoints = 0;
  let goalBonusPoints = 0;

  const isExactScore =
    prediction.predictedHomeScore === match.homeScore &&
    prediction.predictedAwayScore === match.awayScore;

  if (isExactScore) {
    exactScorePoints = 3;
  } else if (
    isCorrectOutcome(
      prediction.predictedHomeScore,
      prediction.predictedAwayScore,
      match.homeScore,
      match.awayScore,
    )
  ) {
    outcomePoints = 1;
  }

  if (
    prediction.predictedTopScorer &&
    match.topScorerName &&
    normalizePlayerName(prediction.predictedTopScorer) ===
      normalizePlayerName(match.topScorerName)
  ) {
    topScorerPoints = 10;
    goalBonusPoints = (match.topScorerGoals ?? 0) * GOAL_BONUS_PER_GOAL;
  }

  return {
    exactScorePoints,
    outcomePoints,
    topScorerPoints,
    goalBonusPoints,
    total: exactScorePoints + outcomePoints + topScorerPoints + goalBonusPoints,
  };
}

File: backend-go/internal/scoring/scoring.go
package scoring

import "strings"

type Prediction struct {
	PredictedHomeScore int
	PredictedAwayScore int
	PredictedTopScorer *string
}

type MatchResult struct {
	HomeScore      int
	AwayScore      int
	TopScorerName  *string
	TopScorerGoals *int
}

type Breakdown struct {
	ExactScorePoints int `json:"exactScorePoints"`
	OutcomePoints    int `json:"outcomePoints"`
	TopScorerPoints  int `json:"topScorerPoints"`
	GoalBonusPoints  int `json:"goalBonusPoints"`
	Total            int `json:"total"`
}

func normalize(s string) string {
	return strings.ToLower(strings.TrimSpace(s))
}

func sign(n int) int {
	if n > 0 {
		return 1
	}
	if n < 0 {
		return -1
	}
	return 0
}

func ScorePrediction(p Prediction, m MatchResult) Breakdown {
	var b Breakdown

	isExact := p.PredictedHomeScore == m.HomeScore && p.PredictedAwayScore == m.AwayScore
	if isExact {
		b.ExactScorePoints = 3
	} else if sign(p.PredictedHomeScore-p.PredictedAwayScore) == sign(m.HomeScore-m.AwayScore) {
		b.OutcomePoints = 1
	}

	if p.PredictedTopScorer != nil && m.TopScorerName != nil &&
		normalize(*p.PredictedTopScorer) == normalize(*m.TopScorerName) {
		b.TopScorerPoints = 10
		if m.TopScorerGoals != nil {
			b.GoalBonusPoints = *m.TopScorerGoals
		}
	}

	b.Total = b.ExactScorePoints + b.OutcomePoints + b.TopScorerPoints + b.GoalBonusPoints
	return b
}
